File: backend/internal/models/booking.go
package models

import (
	"time"
)

// Booking 预订模型
// 对应数据库中的 bookings 表
type Booking struct {
	ID             uint      `gorm:"primaryKey" json:"id"`                                 // 主键
	BookingNumber  string    `gorm:"unique;not null;size:50;index" json:"booking_number"`  // 预订单号（唯一）
	UserID         uint      `gorm:"not null;index" json:"user_id"`                        // 用户 ID（有索引）
	RoomID         uint      `gorm:"not null;index" json:"room_id"`                        // 房间 ID（有索引）
	CheckIn        time.Time `gorm:"not null;index" json:"check_in"`                       // 入住日期（有索引）
	CheckOut       time.Time `gorm:"not null;index" json:"check_out"`                      // 退房日期（有索引）
	TotalDays      int       `gorm:"not null" json:"total_days"`                           // 总天数
	TotalPrice     float64   `gorm:"not null;type:decimal(10,2)" json:"total_price"`       // 总价
	GuestName      string    `gorm:"not null;size:50" json:"guest_name"`                   // 入住人姓名
	GuestPhone     string    `gorm:"not null;size:20" json:"guest_phone"`                  // 入住人电话
	GuestIDCard    string    `gorm:"size:50" json:"guest_id_card"`                         // 入住人身份证号
	SpecialRequest string    `gorm:"type:text" json:"special_request"`                     // 特殊要求
	Status         string    `gorm:"default:'pending';size:20;index" json:"status"`        // 状态：pending, confirmed, checkin, checkout, cancelled
	PaymentStatus  string    `gorm:"default:'unpaid';size:20;index" json:"payment_status"` // 支付状态：unpaid, paid, refunded
	PaymentMethod  string    `gorm:"size:50" json:"payment_method"`                        // 支付方式：wechat, alipay, card
	CancelReason   string    `gorm:"type:text" json:"cancel_reason"`                       // 取消原因
	CreatedAt      time.Time `json:"created_at"`                                           // 创建时间
	UpdatedAt      time.Time `json:"updated_at"`                                           // 更新时间

	// 关联查询（可选）
	// 当查询 Booking 时，可以同时加载 User 和 Room 的信息
	User User `gorm:"foreignKey:UserID" json:"user,omitempty"` // 关联的用户
	Room Room `gorm:"foreignKey:RoomID" json:"room,omitempty"` // 关联的房间
}

// TableName 指定表名
func (Booking) TableName() string {
	return "bookings"
}

// IsPending 判断是否待确认
func (b *Booking) IsPending() bool {
	return b.Status == "pending"
}

// IsConfirmed 判断是否已确认
func (b *Booking) IsConfirmed() bool {
	return b.Status == "confirmed"
}

// IsCancelled 判断是否已取消
func (b *Booking) IsCancelled() bool {
	return b.Status == "cancelled"
}

// IsPaid 判断是否已支付
func (b *Booking) IsPaid() bool {
	return b.PaymentStatus == "paid"
}

// CanCancel 判断是否可以取消
// 只有待确认和已确认的订单可以取消
func (b *Booking) CanCancel() bool {
	return b.Status == "pending" || b.Status == "confirmed"
}

// CanCheckIn 判断是否可以入住
// 已确认且已支付的订单可以入住
func (b *Booking) CanCheckIn() bool {
	return b.Status == "confirmed" && b.IsPaid()
}

File: uniapp/api/booking.js
/**
 * 预订相关 API（按当前后端真实能力适配）
 */

import { get, post } from '@/utils/request.js'

const STATUS_TEXT_MAP = {
  pending: '待确认',
  confirmed: '待入住',
  checkin: '入住中',
  checkout: '已完成',
  cancelled: '已取消',
}

const STATUS_CLASS_MAP = {
  pending: 'pending',
  confirmed: 'confirmed',
  checkin: 'checkin',
  checkout: 'checkout',
  cancelled: 'cancelled',
}

const ensureArray = (value) => {
  if (Array.isArray(value)) {
    return value.filter(Boolean)
  }

  if (typeof value === 'string' && value.trim()) {
    try {
      const parsed = JSON.parse(value)
      return Array.isArray(parsed) ? parsed.filter(Boolean) : []
    } catch (_error) {
      return value
        .split(',')
        .map((item) => item.trim())
        .filter(Boolean)
    }
  }

  return []
}

const ensureNumber = (value, fallback = 0) => {
  const num = Number(value)
  return Number.isFinite(num) ? num : fallback
}

const formatDate = (value) => {
  if (!value) return ''
  return String(value).slice(0, 10)
}

const extractRoomImage = (room = {}) => {
  const images = ensureArray(room.images)
  return images[0] || room.image_url || room.image || 'https://dummyimage.com/720x420/f5f5f5/999999&text=Room'
}

const normalizeBooking = (booking = {}) => {
  const room = booking.room || {}
  const status = booking.status || 'pending'
  const roomId = booking.room_id || room.id || ''
  const roomImage = extractRoomImage(room)

  return {
    id: String(booking.id || ''),
    orderNo: String(booking.booking_number || booking.id || ''),
    status,
    statusText: STATUS_TEXT_MAP[status] || status,
    statusClass: STATUS_CLASS_MAP[status] || 'default',
    roomId: String(roomId),
    roomName: room.room_type || room.name || '房型待确认',
    roomType: room.room_type || room.name || '房型待确认',
    roomNumber: room.room_number || '',
    roomImage,
    hotelName: '七天酒店',
    hotelImage: roomImage,
    checkIn: formatDate(booking.check_in),
    checkOut: formatDate(booking.check_out),
    totalDays: ensureNumber(booking.total_days, 1),
    totalPrice: ensureNumber(booking.actual_price ?? booking.total_price),
    createdAt: booking.created_at || '',
    guestName: booking.guest_name || '',
    guestPhone: booking.guest_phone || '',
    guestIdCard: booking.guest_id_card || '',
    specialRequest: booking.special_request || '',
    paymentStatus: booking.payment_status || 'unpaid',
    canCancel: ['pending', 'confirmed'].includes(status),
    raw: booking,
  }
}

export const createBooking = (data) => {
  const payload = {
    room_id: Number(data.room_id || data.roomId),
    check_in: data.check_in || data.checkIn,
    check_out: data.check_out || data.checkOut,
    guest_name: data.guest_name || data.guestName,
    guest_phone: data.guest_phone || data.guestPhone,
    guest_id_card: data.guest_id_card || data.guestIdCard || '',
    special_request: data.special_request || data.specialRequest || '',
  }

  return post('/bookings', payload)
}

export const getBookingList = async (params = {}) => {
  const result = await get('/bookings/my', {
    page: params.page || 1,
    page_size: params.pageSize || params.page_size || 50,
  })

  const list = (Array.isArray(result) ? result : []).map(normalizeBooking)

  if (!params.status || params.status === 'all') {
    return list
  }

  return list.filter((item) => item.status === params.status)
}

export const getBookingDetail = async (id) => {
  const result = await get(`/bookings/${id}`)
  return normalizeBooking(result)
}

export const cancelBooking = (id, reason = '') => {
  return post(`/bookings/${id}/cancel`, { reason })
}

export const confirmBooking = async () => {
  throw new Error('当前版本暂不支持住客端确认预订')
}

export const calculatePrice = async () => {
  throw new Error('当前版本暂不支持住客端预估价格接口')
}

export const payBooking = async () => {
  throw new Error('当前版本暂不支持住客端支付')
}
